File: csrc/102.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> levels;
        if (root == NULL) return levels;
        queue<TreeNode*> nodeQueues[2];
        nodeQueues[0].push(root);
        int level = 0;
        while (!nodeQueues[level].empty()) {
            vector<int> numsInLevel;
            level = 1 - level;
            while (!nodeQueues[1 - level].empty()) {
                TreeNode* node = nodeQueues[1 - level].front();
                nodeQueues[1 - level].pop();
                numsInLevel.push_back(node->val);
                if (node->left != NULL) nodeQueues[level].push(node->left);
                if (node->right != NULL) nodeQueues[level].push(node->right);
            }
            levels.push_back(numsInLevel);
        }
        return levels;
    }
};

int main() {
    Solution solution;
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    auto ans = solution.levelOrder(root);
    for (auto it = ans.begin(); it != ans.end(); it++) {
        for (auto it2 = (*it).begin(); it2 != (*it).end(); it2++)
            cout << *it2 << " ";
        cout << endl;
    }
    return 0;
}

File: pysrc/102.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return []
        ans = []
        currDepth = 0
        layer = []
        queue = [(root, 0)]
        while len(queue) > 0:
            node, depth = queue.pop(0)
            if node is not None:
                if currDepth != depth:
                    if len(layer) > 0:
                        ans.append(layer)
                        layer = []
                    currDepth = depth
                layer.append(node.val)
                queue.append((node.left, depth + 1))
                queue.append((node.right, depth + 1))
        if len(layer) > 0:
            ans.append(layer)
        return ans


if __name__ == '__main__':
    solution = Solution()
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print(solution.levelOrder(root))
